### PR TITLE
Gh2172

### PR DESCRIPTION
**Pull request by mostlyfabulous**
## Purpose
Extend the `Selector` API to have a `shadowRoot` function that returns the elements inside the `shadowRoot` when the shadow DOM is `open`. These elements are only otherwise accessible with the `shadowRoot` property when using statements like `document.querySelector('css-selector').shadowRoot.querySelectorAll('css-selector');` in a `Selector`, but this creates `Selector` chaining problems if there are child/nested elements with a `shadowRoot` on them as well.

e.g.

`Selector('elem-with-shadowRoot').child('p').shadowRoot.withText('text');` will not work.

This PR should let developers chain existing `Selector` statements that target elements in the `shadowRoot`.

## Approach
Add a new API `shadowRoot` that can be applied to a `Selector` which returns the contents of an element's `shadowRoot` if it exists and `null` otherwise.

## References
#2172

https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM

## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [ ] Add API tests
- [x] Add TS definition
- [ ] Make sure that existing tests do not fail

### CODE REVIEW

**Comment by AlexKamaev** [comment]
@mostlyfabulous 
Thank you for your contribution.
As I understand at this moment the PR is not ready yet. I tried the following test and it does not work as expected:
```JS
import { Selector } from 'testcafe';

fixture `Selector`
    .page `../pages/index.html`;

test('shadowRoot selector API tests', async t => {
    await t.click(Selector('#hasShadowDOM').shadowRoot().child('p').withText('Text should be found!'));
});
```
I debugged your code and found a couple of issues to fix.
I added some modifications to your code to make the test work. Please take a look at them: https://github.com/DevExpress/testcafe/pull/5560/commits/0e9e4cfe403cda8ccbd88c9ea0e98d62690ca61f

We would appreciate it if you review the recommendations from the following comment: https://github.com/DevExpress/testcafe/issues/2172#issuecomment-696031515

**Comment by mostlyfabulous** [comment]
## IE11 does not support `attachShadow` and must be polyfilled. Would prefer to skip IE11 tests for `shadowRoot` API method:
image

image

## Unrelated PR build errors:
image

image

**Comment by AlexKamaev** [comment]
@mostlyfabulous 
I agree, that we need just to skip the tests in IE11, since it does not support the Shadow API. You can skip the test using the `skip` property instead of using the `only` property as here:
https://github.com/DevExpress/testcafe/blob/d7918a1fa4e6a5f76498acd5da1069e55a5c2618/test/functional/fixtures/regression/gh-1366/test.js#L3

 

Let remove all IE related code in your test, such as the follows:
```JS
const isIEFunction = ClientFunction(() => {
```
In addition, there is no need to make/copy test for all selector properties. Test with one property is enough

**Comment by mostlyfabulous** [comment]
>Then, remove all IE related code in your test as follows:

Do you mean to wrap any code specific to IE in the function you provided?

**Comment by AlexKamaev** [comment]
> > Then, remove all IE related code in your test as follows:
> 
> Do you mean to say, wrap any code specific to IE in the function you provided?

No. I saw this function in your code. Probably you accidentally copied it from another test file. I mean, that we do not need any additional code for IE, since IE does not support shadow DOM.
The only thing that we need to do with IE for Shadow DOM tests - is only skipping IE in our test suites. Thus, this `isIEFunction` should be deleted.

**Comment by mostlyfabulous** [comment]
@testcafe-build-bot retest

**Comment by mostlyfabulous** [comment]
@testcafe-build-bot restart

**Comment by AlexKamaev** [comment]
@mostlyfabulous
I see the failed tests. This happens because the `/fixtures/api/es-next/selector/pages/index.html` page is used for Chrome, Firefox, and IE. So, I don't recommend adding your changes to the `/fixtures/api/es-next/selector/pages/index.html` page. Please create a separate page for your test.

Please also rebase your commit to the DevExpress/master branch since some npm packages need to be refreshed.

**Comment by AlexKamaev** [comment]
@mostlyfabulous
Hello. The PR is almost ready. Do you need any assistance to complete it? If you have any difficulties, feel free to ask questions here.

**Comment by mostlyfabulous** [comment]
I should be able to get round to the changes you suggested this week **but feel free** to finish the outstanding changes so this can be released!

**Comment by AlexKamaev** [comment]
Ok,
If I complete my current tasks before you're able to continue working this PR, I'll finish it.

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

File: src/client/driver/command-executors/client-functions/selector-executor/index.js [hunk 1]
```diff
@@ -63,6 +64,9 @@ export default class SelectorExecutor extends ClientFunctionExecutor {
                 const isElementVisible   = !this.command.visibilityCheck || visible(el);
                 const isTimeout          = new DateCtor() - startTime >= this.timeout;
 
+                if (domUtils.isShadowRoot(el))
+                    throw new ActionElementIsShadowRootElementError();
```
Comment (AndreyBelym) [inline src/client/driver/command-executors/client-functions/selector-executor/index.js:68]: I don't feel that here is the right place for this check. By doing it here, we prevent using expressions like `t.expect(element.shadowRoot()).contains('text')` that are perfectly valid in my opinion. I think it's better to do it here:
https://github.com/DevExpress/testcafe/blob/master/src/client/driver/command-executors/execute-action.js#L157

**Review by AndreyBelym — COMMENTED** [review]
An excellent PR, thank you. Just some moments to discuss :)

File: src/client-functions/selectors/add-api.js [hunk 2]
```diff
@@ -25,8 +25,8 @@ const filterNodes = (new ClientFunctionBuilder((nodes, filter, querySelectorRoot
     const result = [];
 
     if (typeof filter === 'string') {
-        // NOTE: we can search for elements only in document or element.
-        if (querySelectorRoot.nodeType !== 1 && querySelectorRoot.nodeType !== 9)
+        // NOTE: we can search for elements only in document/element/shadow root.
+        if (querySelectorRoot.nodeType !== 1 && querySelectorRoot.nodeType !== 9 && querySelectorRoot.toString() !== '[object ShadowRoot]')
```
[the remark this answers is not in the document]
Comment (AlexKamaev) [reply src/client-functions/selectors/add-api.js:29]: That was my initial approach, but then I saw the hammerhead implementation =)

File: src/client-functions/selectors/add-api.js [hunk 3]
```diff
@@ -762,6 +762,28 @@ function addHierarchicalSelectors (options) {
 
         return createDerivativeSelectorWithFilter(args);
     };
+
+    // ShadowRoot
+    obj.shadowRoot = (filter, dependencies) => {
```
[the remark this answers is not in the document]
Comment (mostlyfabulous) [reply src/client-functions/selectors/add-api.js:767]: Makes sense to me now

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Comment by AlexKamaev** [comment]
@AndreyBelym FPR

**Review by AndreyBelym — APPROVED** [review]

**Pull request merged**